sdk/identity/azure-identity/src/azure_pipelines_credential.cpp

```diff
@@ -0,0 +1,258 @@
+
+    throw AuthenticationException(message);
+  }
+
+  const std::string oidcTokenPropertyName = "oidcToken";
+  if (!parsedJson.contains(oidcTokenPropertyName) || !parsedJson[oidcTokenPropertyName].is_string())
+  {
+    std::string message = GetCredentialName()
+        + " : OIDC token not found in response. \nSee Azure::Core::Diagnostics::Logger for details "
+          "(https://aka.ms/azsdk/cpp/identity/troubleshooting).";
+    IdentityLog::Write(IdentityLog::Level::Verbose, message);
+    throw AuthenticationException(message);
+  }
+  return parsedJson[oidcTokenPropertyName].get<std::string>();
+}
+
+AzurePipelinesCredential::~AzurePipelinesCredential() = default;
+
+AccessToken AzurePipelinesCredential::GetToken(
+    TokenRequestContext const& tokenRequestContext,
+    Context const& context) const
+{
+  if (!m_tokenCredentialImpl)
+  {
+    auto const AuthUnavailable = GetCredentialName() + " authentication unavailable. ";
+
+    IdentityLog::Write(
+        IdentityLog::Level::Warning,
+        AuthUnavailable + "See earlier " + GetCredentialName() + " log messages for details.");
+
+    throw AuthenticationException(
+        AuthUnavailable + "Azure Pipelines environment is not set up correctly.");
+  }
+
+  auto const tenantId = TenantIdResolver::Resolve(
+      m_clientCredentialCore.GetTenantId(),
+      tokenRequestContext,
+      m_clientCredentialCore.GetAdditionallyAllowedTenants());
+
+  auto const scopesStr
+      = m_clientCredentialCore.GetScopesString(tenantId, tokenRequestContext.Scopes);
+
+  // TokenCache::GetToken() and m_tokenCredentialImpl->GetToken() can only use the lambda
+  // argument when they are being executed. They are not supposed to keep a reference to lambda
+  // argument to call it later. Therefore, any capture made here will outlive the possible time
+  // frame when the lambda might get called.
+  return m_tokenCache.GetToken(scopesStr, tenantId, tokenRequestContext.MinimumExpiration, [&]() {
+    return m_tokenCredentialImpl->GetToken(context, false, [&]() {
+      auto body = m_requestBody;
+      if (!scopesStr.empty())
+      {
+        body += "&scope=" + scopesStr;
+      }
+
+      auto const requestUrl = m_clientCredentialCore.GetRequestUrl(tenantId);
+
+      Azure::Core::Http::Request oidcRequest = CreateOidcRequestMessage();
+      std::unique_ptr<RawResponse> response = m_httpPipeline.Send(oidcRequest, context);
+
+      if (!response)
+      {
+        throw AuthenticationException(
+            GetCredentialName() + " couldn't send OIDC token request: null response.");
+      }
+
+      auto const bodyStream = response->ExtractBodyStream();
+      auto const bodyVec = bodyStream ? bodyStream->ReadToEnd(context) : response->GetBody();
+      auto const responseBody
+          = std::string(reinterpret_cast<char const*>(bodyVec.data()), bodyVec.size());
+
+      std::string assertion = GetOidcTokenResponse(response, responseBody);
```

Comment:
nit. To my taste, I'd move all the stuff related to oidc request to the very top of the callback, otherwise it looks a bit strange that we calculate `requestUrl` and other stuff before sending an OIDC request, and I read and expect to see the stuff above to be used below, in the OIDC request, but it is not happening, so it sits there until line 251 when it gets used. I'd move these calculations closer to the place where they are used.

```suggestion
      Azure::Core::Http::Request oidcRequest = CreateOidcRequestMessage();
      std::unique_ptr<RawResponse> response = m_httpPipeline.Send(oidcRequest, context);

      if (!response)
      {
        throw AuthenticationException(
            GetCredentialName() + " couldn't send OIDC token request: null response.");
      }

      auto const bodyStream = response->ExtractBodyStream();
      auto const bodyVec = bodyStream ? bodyStream->ReadToEnd(context) : response->GetBody();
      auto const responseBody
          = std::string(reinterpret_cast<char const*>(bodyVec.data()), bodyVec.size());

      const std::string assertion = GetOidcTokenResponse(response, responseBody);

      auto body = m_requestBody;
      if (!scopesStr.empty())
      {
        body += "&scope=" + scopesStr;
      }

      auto const requestUrl = m_clientCredentialCore.GetRequestUrl(tenantId);
```

Reply:
BTW, what if you move all that OIDC request logic to the separate function, something like `std::string GetAssertion()`?

Reply:
That's the plan. With ClientAssertionCredential, this whole logic moves into an assertion callback. So, will refactor that after.

Reply:
> To my taste, I'd move all the stuff related to oidc request to the very top of the callback

I agree, it looks nicer, but...
The main reason why I didn't have it at the top was because I wanted the AuthorityHost https check to happen **first** (and potentially throw), before sending the oidc request.

I can add a comment to explain the rationale.

This line validates it
```cpp
auto const requestUrl = m_clientCredentialCore.GetRequestUrl(tenantId);
```

And throws:
```cpp
if (m_authorityHost.GetScheme() != "https")
{
  throw Azure::Core::Credentials::AuthenticationException(
      "Authority host must be a TLS protected (https) endpoint.");
}
```